### ISSUE TITLE
Class name of TypeParameter

### ISSUE THREAD

**Issue by mlangkabel** [ftomassetti/java-symbol-solver#90]
@ftomassetti: While you're in refactoring mode, I'd like to bring up a naming issue I've seen in the SymbolSolver codebase: Type parameters.

Sometimes they are calles TypeParameter, sometimes just parameter, sometimes TypeVariable and I'm not totally sure but maybe I've also seen some code calling them actual and formal parameters. 
I think we should consistently stick to one path and call it `TypeParameter` and `TypeArgument` or call it `FormalTypeParameter` and `ActualTypeParameter`. 

I've been working with Clang a lot over the last three years and they call it parameter and argument and I think the Javaparser also uses this convention. What do you think?

**Comment by ftomassetti** [comment]
Yes, this is a very good point. I get confused myself.

Looking at [JLS 4.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.4):

> A type variable is an unqualified identifier used as a type in class, interface, method, and constructor bodies.

> A type variable is introduced by the declaration of a type parameter of a generic class, interface, method, or constructor (§8.1.2, §9.1.2, §8.4.4, §8.8.4).

Applying these to classes we could have:
* `me.tomassetti.symbolsolver.model.declarations.TypeParameterDeclaration` (all these things are called declaration)
* `me.tomassetti.symbolsolver.model.usages.typesystem.TypeVariable` (without the `Type` at the end because it would be too long)

Does it make sense?
If yes, we could start renaming these classes and then look at the specific names used fields and methods, around the code base.
What do you think?

**Comment by mlangkabel** [comment]
So a TypeVariable is a TypeParameterUsage? 

And what about TypeArguments? In my understanding a class declaration may have type parameters. But a type usage of that type has type arguments. 
The same should be true for methods. A method has parameters but the method usage has arguments.

I'd strongly suggest that we use the terminology like that. Otherwise it's getting confusing quickly ;)

**Comment by ftomassetti** [comment]
Ok, so you mean using _parameter_ for the formal parameter (the declaration) and argument for the actual values (the usage). Ok, we could do that.

**Comment by mlangkabel** [comment]
Exactly. Here some examples of how I'm understanding what we are planning to do:
```
public class Foo<T> {		// T is a type parameter
	public void bar1(int i) {	// i is parameter
		int k = i;			// i is an argument
	}

	private T t;			// T is a type parameter usage
}

public class Bar<T2>{		// T2 is a type parameter
	private Foo<T2> f;		// T2 is a type parameter usage used as a type argument
}
```

**Comment by ftomassetti** [comment]
There is this interface:

```
public interface TypeParametrizable {

    List<TypeParameterDeclaration> getTypeParameters();

}
```

so if I have a class declaration I can call `getTypeParameters` and I get for example `[T]`

in `private T t;` `T` would be a `TypeVariable` (I am using the JLS terminology here.

**Comment by ftomassetti** [comment]
`MethodLikeDeclaration` (an ancestor for `ConstructorDeclaration` and `MethodDeclaration`) has the following methods:

```
    int getNoParams();

    ParameterDeclaration getParam(int i);

    /**
     * The last parameter can be variadic and sometimes it needs to be handled in a special way.
     */
    default ParameterDeclaration getLastParam() {
        if (getNoParams() == 0) {
            throw new UnsupportedOperationException("This method has no typeParametersValues, therefore it has no a last parameter");
        }
        return getParam(getNoParams() - 1);
    }

```

Does it make sense?

**Comment by ftomassetti** [comment]
However `Context` has these methods:
```
SymbolReference<MethodDeclaration> solveMethod(String name, List<Type> argumentsTypes, TypeSolver typeSolver);
Optional<MethodUsage> solveMethodAsUsage(String name, List<Type> argumentsTypes, TypeSolver typeSolver)
```

Here we use the term `argumentsTypes` because they would be the types of the actual values passed

**Comment by mlangkabel** [comment]
That sounds great to me, I just have one or two small suggestion to make:
* We should add a comment to the TypeVariable class that explains what that is (for people like me who only know the argument/parameter distinction)
* We should use the term parameter in the context of a `MethodDeclaration` (so change the code in the example one comment above this one), and use the term argument when we refer to some `MethodUsage`. I think that would make sense :)

**Comment by ftomassetti** [comment]
The first point is addressed. About the second one I renamed parametersTypes to argumentTypes in a bunch of places. Did I miss anything?

**Comment by mlangkabel** [comment]
Just posted a comment on that pull request.
I think I just realized some of my earlier confusion: I thought a MethodUsage would be something like a call or a reference to a method. But in fact a MethodUsage is just a MethodDeclaration with TypeArguments replacing the TypeParameters! So I have to revert what I said before: Don't use the term "arguments" in the context of a MethodUsage. Sorry for the confusion.

**Comment by mlangkabel** [comment]
Oh, and you renamed TypeUsage to Type (so it does not feel like a strange third thing (next to Declaration and Type) anymore) and TypeParameterUsage to TypeVariable. Awesome!, that's what I had in mind, too :)

**Comment by ftomassetti** [comment]
Thanks for your help on this one!
Closed by #92 but I expect we will have to fix some names here and there. Please feel free to discuss naming because poor names do not help in this already complex project.
